### PR TITLE
[ISSUE #3925] Init retry and factory code in sdk module.

### PR DESCRIPTION
**Pull request by ShawnJim**
ref #3925 .

## Task List
- [X] shenyu-sdk-retry.
- [X] shenyu-sdk-factory.

## To Be Continue
- [ ] append loadbalancer module.
- [ ] unit test.

### CODE REVIEW

File: shenyu-sdk/shenyu-sdk-core/src/main/java/org/apache/shenyu/sdk/core/ShenyuRequest.java [hunk 1]
```diff
@@ -93,6 +93,18 @@ public static ShenyuRequest create(final HttpMethod httpMethod,
...
     }
 
+    /**
+     * Builds a Request. All parameters must be effectively immutable, via safe copies.
+     *
+     * @param url for the request.
+     * @param request to include.
+     * @return a Request
+     */
+    public static ShenyuRequest create(final String url,
```
Comment (moremind) [inline shenyu-sdk/shenyu-sdk-core/src/main/java/org/apache/shenyu/sdk/core/ShenyuRequest.java:103]: Why the line break

File: shenyu-sdk/shenyu-sdk-core/src/main/java/org/apache/shenyu/sdk/core/http/AbstractShenyuHttpClient.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    private static final String URL_REWRITE_REGEX = "http:\\/\\/[a-z\\d\\.|:]+\\/";
+
+    private final Retryer retryer;
+
+    private final ShenyuInstanceRegisterRepository registerRepository;
+
+    private final ShenyuConfig.RegisterConfig sdkConfig;
+
+    public AbstractShenyuHttpClient() {
+        this.sdkConfig = Optional.ofNullable(Singleton.INST.get(ShenyuConfig.class)).orElse(new ShenyuConfig()).getSdk();
```
Comment (moremind) [inline shenyu-sdk/shenyu-sdk-core/src/main/java/org/apache/shenyu/sdk/core/http/AbstractShenyuHttpClient.java:50]: this parameter always nonnull,you should just shenyuConfig.sdk

File: shenyu-sdk/shenyu-sdk-core/src/main/java/org/apache/shenyu/sdk/core/http/AbstractShenyuHttpClient.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+     * TODO wait append load balance logic.
+     * get rewrite url by request.
+     *
+     * @param request the request.
+     * @return {@linkplain String}
+     */
+    private String getRewriteUrl(final ShenyuRequest request) {
+        String url;
+
+        if (!sdkConfig.getRegisterType().equals("local")) {
```
Comment (moremind) [inline shenyu-sdk/shenyu-sdk-core/src/main/java/org/apache/shenyu/sdk/core/http/AbstractShenyuHttpClient.java:89]: this code is error,maybe npe.

File: shenyu-sdk/shenyu-sdk-core/src/main/java/org/apache/shenyu/sdk/core/http/RetryableException.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+import java.util.Date;
+
+/**
+ * This exception is raised when the {@link org.apache.shenyu.sdk.core.ShenyuResponse} is deemed to be retryable, typically via an
+ * {@link org.apache.shenyu.common.exception.CommonErrorCode} when the {@link ShenyuResponse#getStatus() status} is 503.
+ * Reference to feign.RetryableException.
+ */
+public class RetryableException extends ShenyuException {
+
+    private static final long serialVersionUID = 1L;
```
Comment (moremind) [inline shenyu-sdk/shenyu-sdk-core/src/main/java/org/apache/shenyu/sdk/core/http/RetryableException.java:34]: not use 1L

File: shenyu-sdk/shenyu-sdk-core/src/main/java/org/apache/shenyu/sdk/core/http/RetryableException.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+     * @param message message
+     * @param cause cause
+     * @param retryAfter usually corresponds to the retryAfter header.
+     * @param request request
+     */
+    public RetryableException(final String message, final java.lang.Throwable cause,
+                     final Date retryAfter, final ShenyuRequest request) {
+        super(message, cause);
+        this.httpMethod = request.getHttpMethod();
+        this.retryAfter = retryAfter != null ? retryAfter.getTime() : null;
```
Comment (moremind) [inline shenyu-sdk/shenyu-sdk-core/src/main/java/org/apache/shenyu/sdk/core/http/RetryableException.java:52]: refactor this code

File: shenyu-sdk/shenyu-sdk-core/src/main/java/org/apache/shenyu/sdk/core/http/Retryer.java [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+    void continueOrPropagate(RetryableException e);
+
+    /**
+     * clone current retryer.
+     *
+     * @return {@linkplain Retryer}
+     */
+    Retryer instance();
+
+    class Default implements Retryer {
```
Comment (moremind) [inline shenyu-sdk/shenyu-sdk-core/src/main/java/org/apache/shenyu/sdk/core/http/Retryer.java:62]: not use default class

**Review by moremind — CHANGES_REQUESTED** [review]

File: shenyu-sdk/shenyu-sdk-core/src/main/java/org/apache/shenyu/sdk/core/ShenyuRequest.java (hunk 1 above)
Comment (ShawnJim) [reply shenyu-sdk/shenyu-sdk-core/src/main/java/org/apache/shenyu/sdk/core/ShenyuRequest.java]: done

File: shenyu-sdk/shenyu-sdk-core/src/main/java/org/apache/shenyu/sdk/core/http/AbstractShenyuHttpClient.java (hunk 2 above)
Comment (ShawnJim) [reply shenyu-sdk/shenyu-sdk-core/src/main/java/org/apache/shenyu/sdk/core/http/AbstractShenyuHttpClient.java]: done

File: shenyu-sdk/shenyu-sdk-core/src/main/java/org/apache/shenyu/sdk/core/http/RetryableException.java (hunk 5 above)
Comment (ShawnJim) [reply shenyu-sdk/shenyu-sdk-core/src/main/java/org/apache/shenyu/sdk/core/http/RetryableException.java]: done

File: shenyu-sdk/shenyu-sdk-core/src/main/java/org/apache/shenyu/sdk/core/http/RetryableException.java (hunk 4 above)
Comment (ShawnJim) [reply shenyu-sdk/shenyu-sdk-core/src/main/java/org/apache/shenyu/sdk/core/http/RetryableException.java]: done

File: shenyu-sdk/shenyu-sdk-core/src/main/java/org/apache/shenyu/sdk/core/http/AbstractShenyuHttpClient.java (hunk 3 above)
Comment (ShawnJim) [reply shenyu-sdk/shenyu-sdk-core/src/main/java/org/apache/shenyu/sdk/core/http/AbstractShenyuHttpClient.java]: done

**Review by moremind — APPROVED** [review]

**Pull request merged**